test/boot_raw_fallback.c

```diff
@@ -0,0 +1,414 @@
+#include <locale.h>
+#include <glib.h>
+#include <gio/gio.h>
+#include <glib/gstdio.h>
+
+#include "update_handler.h"
+#include "common.h"
+#include "context.h"
+
+typedef struct {
+	gchar *tmpdir;
+} BootRawFallbackFixture;
+
+#define BIT(nr) (1UL << (nr))
+typedef enum
+{
+	OPT_NONE                      = 0,
+	OPT_EXPECT_FAIL               = BIT(0),
+	OPT_PRE_INSTALL_PRIMARY_HEAD  = BIT(1),
+	OPT_PRE_INSTALL_PRIMARY_IMG   = BIT(2),
+	OPT_PRE_INSTALL_FALLBACK_HEAD = BIT(3),
+	OPT_PRE_INSTALL_FALLBACK_IMG  = BIT(4),
+	OPT_DUMMY_FILL_PRIMARY        = BIT(5),
+	OPT_DUMMY_FILL_FALLBACK       = BIT(6),
+} BootRawFallbackOptions;
+
+typedef struct {
+	BootRawFallbackOptions options;
+
+	GQuark err_domain;
+	gint err_code;
+
+	gsize image_size;
+	guint64 region_start;
+	guint64 region_size;
+} BootRawFallbackData;
+
+static gboolean process_file(const gchar *filepath,
+		gsize offset,
+		gsize size,
+		gchar fill,
+		gboolean random,
+		gboolean compare)
+{
+	GError *error = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GFileIOStream) stream = NULL;
+	g_autofree gchar *buf = NULL;
+	g_autofree gchar *cmp_buf = NULL;
+	const gsize buf_size = 4096;
+	gboolean res;
+
+	file = g_file_new_for_path(filepath);
+	stream = g_file_open_readwrite(file, NULL, &error);
+	g_assert_no_error(error);
+	g_assert_nonnull(stream);
+
+	res = g_seekable_seek(G_SEEKABLE(stream), offset, G_SEEK_SET, NULL, &error);
+	g_assert_no_error(error);
+	g_assert_true(res);
+
+	if (!random) {
+		buf = g_malloc(buf_size);
+		memset(buf, fill, buf_size);
+	}
+
+	if (compare)
+		cmp_buf = g_malloc(buf_size);
+
+	while (size) {
+		gsize bytes_to_process = size < buf_size ? size : buf_size;
+		gsize bytes_processed;
+
+		if (random)
+			buf = (gchar*) random_bytes(bytes_to_process, 31415);
```

Comment:
This leaks memory on every loop. It would probably be simpler (and avoid repeating patterns) to either use g_malloc+memset or random_bytes for the full size.

Reply:
As these partitions aren't multiple MB in size, one large allocation wouldn't hurt, but it's fine this way as well.